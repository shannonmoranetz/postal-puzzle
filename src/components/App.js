import React, { Component } from 'react';
import AnswerBank from './AnswerBank';
import '../../src/styles/main.scss';

export default class App extends Component {
  constructor() {
    super();
    this.state = {
      allQuestions: [],
      isLoaded: false,
      currentQuestionCount: 0,
      score: 0,
      answerIsCorrect: true
    };
  };

  componentDidMount = () => {
    this.fetchData()
  };

  fetchData = () => {
    fetch('http://memoize-datasets.herokuapp.com/api/v1/questionData')
    .then(response => response.json())
    .then(data => {
      this.setState({
        allQuestions: data.questionData,
        isLoaded: true,
        score: 0
      })
    })
    .catch(error => console.log(error));
  }

  shuffleAnswers = () => {
    let shuffledArray = [this.state.allQuestions[this.state.currentQuestionCount].correctAnswer]
    shuffledArray = shuffledArray.concat(this.state.allQuestions[this.state.currentQuestionCount].incorrectAnswers)
    shuffledArray.sort(() => {
      return 0.5 - Math.random()
    })
    return shuffledArray;
  }

  checkAnswer = (isCorrect) => {
    if (isCorrect === this.state.allQuestions[this.state.currentQuestionCount].correctAnswer) {
      this.updateScoreSum()
      this.setState({
        answerIsCorrect: true
      })
    } else {
      this.setState({
        answerIsCorrect: false
      })
    }
    this.updateCurrentQuestion();
  }

  updateScoreSum = () => {
    let updatedScore = this.state.score + this.state.allQuestions[this.state.currentQuestionCount].pointsValue;
    this.setState({
      score: updatedScore
    })
  }

  updateCurrentQuestion = () => {
    this.setState({
      currentQuestionCount: (this.state.currentQuestionCount + 1)
    })
    this.checkIfGameOver()
  }

  moveToNextQuestion = () => {
    this.setState({
      answerIsCorrect: true
    })
  }

  checkIfGameOver = () => {
    if (this.state.currentQuestionCount === 29) {
      this.setState({
        currentQuestionCount: 0,
        score: 0
      })
    }
  }

  render() {
    if (this.state.isLoaded === true && this.state.answerIsCorrect === true) {
      return (
        <div className="app-container">
         <div className="header-bar">
            <h1 className="app-title">Postal Puzzle</h1>
          </div>
          <main className="main-content">
          <div className="info-container">
            <div className="score-container">
              <p className="score">Score: {this.state.score}</p>
            </div>
            <p className="category"><span className="category-bold">Category: </span>{this.state.allQuestions[this.state.currentQuestionCount].category}</p>
            <p className="difficulty"><span className="difficulty-bold">Difficulty: </span>{this.state.allQuestions[this.state.currentQuestionCount].difficulty}</p>
            <p className="points-value">This question is worth <span className="points-value-bold">{this.state.allQuestions[this.state.currentQuestionCount].pointsValue}</span> points.</p>
          </div>
         <div className="game-container">
            <h3 className="question-count">Question #{(this.state.currentQuestionCount + 1)}</h3>
            <p className="question">{this.state.allQuestions[this.state.currentQuestionCount].question}</p>
            <AnswerBank shuffledAnswers={this.shuffleAnswers()}
                        isLoaded={this.state.isLoaded}
                        checkAnswer={this.checkAnswer}/>
            </div>
          </main>
        </div>
      )
    } else if (this.state.isLoaded === true && this.state.answerIsCorrect === false) {
      return (
        <div className="app-container-alt">
          <div className="header-bar">
            <h1 className="app-title">Postal Puzzle</h1>
          </div>
          <div className="incorrect-info-container">
            <div className="incorrect-info-elements">
              <div className="question-heading">
                <h3 className="question-alt">Question #{(this.state.currentQuestionCount)}</h3>
              </div>
              <p className="incorrect-display">Incorrect...</p>
              <p className="correct-display">The correct answer was:</p>
              <p className="show-correct">{this.state.allQuestions[this.state.currentQuestionCount].correctAnswer}</p>
              <button className="confirm-move-on" onClick={this.moveToNextQuestion}>Next Question</button>
            </div>
          </div>
        </div>
      )
    } else {
      return (
        <div>
          <p className="loading">Now Loading...</p>
        </div>
      )
    }
  };
};

