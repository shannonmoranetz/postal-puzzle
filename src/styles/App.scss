body {
  background-color: pink;
  font-family: $body-font;
}