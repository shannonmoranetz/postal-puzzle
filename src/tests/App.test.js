import React from 'react';
import ReactDOM from 'react-dom';
import App from '../../src/components/App.js';
import { shallow } from 'enzyme';

const currentQuestion = [{
  question: "What is a package?",
  correctAnswer: "One or more modules grouped together that solve a problem.",
  incorrectAnswers: [
  "An app used in React to create components.",
  "A feature specific to Node.js.",
  "Code that can't be re-used once it has executed."
  ],
  category: "Concepts",
  difficulty: "Easy",
  pointsValue: 100,
  questionID: 0
}];

const moveToNextQuestionMock = jest.fn();

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <App moveToNextQuestion={moveToNextQuestionMock}/>
    );
  });

  it('should have the proper default state', () => {
    expect(wrapper.state()).toEqual({
      allQuestions: [],
      isLoaded: false,
      currentQuestionCount: 0,
      score: 0,
      answerIsCorrect: true
    })
  })

  it('should properly render the component elements', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <App moveToNextQuestion={moveToNextQuestionMock}/>
    );
  });

  // it('should call moveToNextQuestion if answerIsCorrect is false', () => {
  //   wrapper.setState({
  //     isLoaded: true,
  //     answerIsCorrect: false,
  //     allQuestions: currentQuestion
  //    })
  //   wrapper.find('.confirm-move-on').simulate('click');
  //   expect(moveToNextQuestionMock).toHaveBeenCalled();
  // });

  it('should only render the QuestionCard and AnswerBank component when loaded', () => {
    wrapper.setState({
      isLoaded: true,
      allQuestions: currentQuestion
     })
    expect(wrapper.find('QuestionCard').length).toEqual(1)
    expect(wrapper.find('AnswerBank').length).toEqual(1)
  })

  it('should check the answer, and if correct, update score and correct marker to true', () => {
    wrapper.setState({
      allQuestions: currentQuestion
    })
    expect(wrapper.state('answerIsCorrect')).toEqual(true);
    wrapper.instance().checkAnswer(currentQuestion[0].correctAnswer);
    expect(wrapper.state('score')).toEqual(100);
    expect(wrapper.state('answerIsCorrect')).toEqual(true);
  });

  it('should check the answer, and if incorrect, update correct marker to false', () => {
    wrapper.setState({
      allQuestions: currentQuestion
    })
    expect(wrapper.state('answerIsCorrect')).toEqual(true);
    wrapper.instance().checkAnswer('Incorrect Answer');
    expect(wrapper.state('answerIsCorrect')).toEqual(false);
  });

  it('should update the score with the current question point value', () => {
    wrapper.setState({ 
      allQuestions: currentQuestion
     })
    expect(wrapper.state('score')).toEqual(0);
    wrapper.instance().updateScoreSum();
    expect(wrapper.state('score')).toEqual(100);
  });

  it('should move to the next question by updating the correct marker to true', () => {
    wrapper.setState({
      answerIsCorrect: false
     })
    expect(wrapper.state('answerIsCorrect')).toEqual(false);
    wrapper.instance().moveToNextQuestion();
    expect(wrapper.state('answerIsCorrect')).toEqual(true);
  });

  it('should only reset the current question count and score to 0 if all questions are answered', () => {
    wrapper.setState({
      currentQuestionCount: 10,
      score: 100
     })
    wrapper.instance().checkIfGameOver();
    expect(wrapper.state('currentQuestionCount')).toEqual(10);
    expect(wrapper.state('score')).toEqual(100);
    wrapper.setState({
      currentQuestionCount: 29
     })
    wrapper.instance().checkIfGameOver();
    expect(wrapper.state('currentQuestionCount')).toEqual(0);
    expect(wrapper.state('score')).toEqual(0);
  });

}); 