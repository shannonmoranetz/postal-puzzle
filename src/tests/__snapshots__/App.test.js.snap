// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should properly render the component elements 1`] = `
<div>
  <p>
    loading...
  </p>
</div>
`;
